README.md
IATI Datastore Classic
======================

[![Build Status](https://github.com/codeforIATI/iati-datastore/workflows/CI/badge.svg?branch=main)](https://github.com/codeforIATI/iati-datastore/actions?query=workflow%3ACI)
[![Coverage Status](https://img.shields.io/coveralls/codeforIATI/iati-datastore.svg)](https://coveralls.io/r/codeforIATI/iati-datastore?branch=main)
[![License: MIT](https://img.shields.io/badge/license-AGPLv3-blue.svg)](https://github.com/codeforIATI/iati-datastore/blob/main/LICENSE.txt)


Introduction
------------

The International Aid Transparency Initiative (IATI) aims to make
information about aid spending easier to access.

IATI maintains the [IATI Standard](https://iatistandard.org) and keeps a
[Registry of IATI data](https://iatiregistry.org/).

The *IATI Datastore* was originally built in 2013 by the
[Open Knowledge Foundation](https://okfn.org).
[Code for IATI](https://codeforiati.org) has updated the
software to use Python3 and modern dependencies.
Over time, we will include other bugfixes and feature
improvements. Our fork of the software is called *IATI Datastore Classic*.

A public instance is available here:

https://datastore.codeforiati.org


Requirements
------------

You will need [Redis](https://redis.io), [Postgres](https://postgresql.org), Python 3, pip and develpment libraries (for libpq, libxml2 and libxslt) to run the full setup.
For example, on Ubuntu:

    sudo apt-get install postgresql redis-server python3 python3-pip libpq-dev libxml2-dev libxslt-dev libevent-dev python3-dev

Installing for development
--------------------------

```
# Clone the source
git clone https://github.com/codeforIATI/iati-datastore.git

# Install development dependencies
pip install -r requirements_dev.txt

# Run the tests  (these tests use an in-memory sqlite db)
nosetests iati_datastore

# Create a new PostgreSQL database
sudo -u postgres psql -c "CREATE DATABASE iati_datastore"

# Set an environment variable for `IATI_DATASTORE_DATABASE_URL` linking to the database created
export IATI_DATASTORE_DATABASE_URL=postgres:///iati_datastore

# Create the db tables
iati db upgrade

# Note: To create the tables the new database may need access privileges granted to your system user
# See http://dba.stackexchange.com/questions/117109/how-to-manage-default-privileges-for-users-on-a-database-vs-schema/117661#117661
sudo -u postgres psql -c "CREATE USER [SYSTEM USER]"
sudo -u postgres psql -c "GRANT ALL ON DATABASE iati_datastore TO [SYSTEM USER]"

# Start the process of grabbing the source data
iati crawler download-and-update

# Start a development server – this should be run in a seperate terminal window
iati run

# Run a worker. This will download and index the datafiles
iati queue background

# The progess of the worker can be checked using:
iati crawler status

# A local API is available at: http://127.0.0.1:5000
```

Deploying with nginx
--------------------

* Intall the requirements listed above
* Install nginx

        sudo apt-get install nginx

* Install uwsgi, from within your `virtualenv`

        pip3 install uwsgi

* Create a uwSGI .ini file inside the root `iati-datastore` folder, e.g.

        [uwsgi]
        module = liveserver
        master = true
        processes = 5
        socket = /var/www/socks/%n.sock
        logto = /var/log/uwsgi/%n.log
        chmod-socket = 666
        vacuum = true
        die-on-term = true

* Make sure `/var/www/socks/` and `/var/log/uwsgi/` are writeable by the `www-data` user.

* Add a `liveserver.py` file, according to your server configuration:

        import sys, os
        PATH = os.path.dirname(os.path.realpath(__file__))
        sys.path.insert(0, '/path/to/iati-datastore/pyenv/lib/python3.6/site-packages')
        sys.path.insert(0, PATH)
        from iati_datastore.iatilib.wsgi import app as application

* Set up a systemd service, e.g. in `/etc/systemd/system/iati-datastore.service`

        [Unit]
        Description=IATI Datastore uWSGI instance
        After=network.target
        [Service]
        User=www-data
        Group=www-data
        WorkingDirectory=/path/to/iati-datastore/
        ExecStart=/path/to/iati-datastore/pyenv/bin/uwsgi --ini /path/to/iati-datastore/iati-datastore_uwsgi.ini
        [Install]
        WantedBy=multi-user.target


* Start the systemd service, e.g.:

        systemctl start iati-datastore

* Make sure that the service loads on restart:

        systemctl enable iati-datastore

* Create your nginx config file, e.g. in /etc/nginx/sites-available/datastore

        server {
            server_name datastore.codeforiati.org;
            gzip            on;
            gzip_types      text/plain application/xml text/css application/javascript application/json;
            gzip_min_length 1000;
            error_log  /var/www/logs/error.log;
            access_log /var/www/logs/access.log;

            error_page 404 /error/404.html;
            error_page 500 501 502 503 504 /error/5xx.html;
            location /error {
                root /path/to/iati-datastore/iati_datastore/iatilib/frontend/templates;
                internal;
            }

            location /docs {
                alias /path/to/iati-datastore/iati_datastore/iatilib/frontend/docs/dirhtml;
                expires 1y;
            }

            location / {
                uwsgi_intercept_errors on;
                include uwsgi_params;
                uwsgi_pass unix:/var/www/socks/iati-datastore_uwsgi.sock;
                uwsgi_buffering off;
            }
        }

* Load the config file and test nginx:

        cp /etc/nginx/sites-available/datastore /etc/nginx/sites-enabled/datastore
        nginx -t

* If everything worked out, restart nginx and your site should be available through your server (e.g. in the above configuration, via `datastore.codeforiati.org`:

        systemctl restart nginx


Deploying with apache
---------------------

* Install the requirements listed above
* Install Apache and mod_wsgi

        sudo apt-get install apache2 libapache2-mod-wsgi

* Clone the source
* Install `pip install -e iati_datastore`
* Create a database (in postgres), and set an environment variable
  `DATABASE_URL`. e.g.:

        sudo -u postgres createdb iati-datastore -O my_username -E utf-8
        export DATABASE_URL='postgres:///iati-datastore'

* Run `iati create_database` to create the db tables
* Set up a cron job for updates. (Add the following line after running `crontab -e`)

        0 0 * * * export DATABASE_URL='postgres:///iati-datastore'; /usr/local/bin/iati crawler download-and-update

* Run a worker with `iati queue background`
    - This needs to persist when you close your ssh connection. A simple way of doing this is using [screen](https://www.gnu.org/software/screen/).

* Set up apache using mod_wsgi

* Create a datastore.wsgi file containing this code (this is necessary because Apache's mod wsgi handles environment variables differently):

        import os
        os.environ['DATABASE_URL'] = 'postgres:///iati-datastore'
        from iatilib.wsgi import app as application

* Add this inside the `<VirtualHost>` tags of your apache configuration:

        WSGIDaemonProcess datastore user=my_username group=my_username
        WSGIProcessGroup datastore
        WSGIScriptAlias / /home/datastore/datastore.wsgi


Updating activities after changing import code
----------------------------------------------

* Run this SQL query on the database - `UPDATE resource SET last_succ=NULL;`
* Restart background process
* Run `iati crawler download-and-update` (or wait for cron to run it for you)


Generation of Documentation
---------------------------

API documentation in the docs folder is generated using [Sphinx](https://www.sphinx-doc.org).

    iati build-docs

Generation of the front page and query builder
----------------------------------------------

    iati build-query-builder

By default, the query builder will look for files served from http://127.0.0.1:5000 - if you would like to point it elsewhere (e.g. to https://datastore.codeforiati.org) then you should add the argument `deploy-url`:

    iati build-query-builder --deploy-url https://datastore.codeforiati.org


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- Get fabric working for deployment ([#113](https://github.com/codeforIATI/iati-datastore/pull/113))
- Recipient region, country and sector also search transactions by default ([#117](https://github.com/codeforIATI/iati-datastore/pull/117))
- Add /api/ and /api/1/ routes ([#120](https://github.com/codeforIATI/iati-datastore/pull/120))
- Add some tests for the generated documentation ([#123](https://github.com/codeforIATI/iati-datastore/pull/123))
- Add some console tests ([#127](https://github.com/codeforIATI/iati-datastore/pull/127))
- Add a home page, including the query builder and information about health status ([#133](https://github.com/codeforIATI/iati-datastore/pull/133))
- Add a route to turn an API request into a populated query builder ([#140](https://github.com/codeforIATI/iati-datastore/pull/140))
- Add search on title and description ([#168](https://github.com/codeforIATI/iati-datastore/pull/168))
- Add a `Stats` model to keep track of totals ([#188](https://github.com/codeforIATI/iati-datastore/pull/188))
- Auto-deploy `main` branch on Github, via Github Actions ([#189](https://github.com/codeforIATI/iati-datastore/pull/189))
- Add issue templates ([#193](https://github.com/codeforIATI/iati-datastore/pull/193))
- Add XML and JSON options to query builder ([#195](https://github.com/codeforIATI/iati-datastore/pull/195))
- Add `iati reset-stats` command ([#199](https://github.com/codeforIATI/iati-datastore/pull/199))
- Add prettier error pages ([#203](https://github.com/codeforIATI/iati-datastore/pull/203))
- Add a changelog ([#222](https://github.com/codeforIATI/iati-datastore/pull/222))

### Changed
- Change xmlns in XML output to https://datastore.codeforiati.org/ns ([#111](https://github.com/codeforIATI/iati-datastore/pull/111))
- Use Flask Migrate for migrations ([#116](https://github.com/codeforIATI/iati-datastore/pull/116))
- Switch to using iatikit to download data much faster and more reliably ([#131](https://github.com/codeforIATI/iati-datastore/pull/131))
- Change default branch from `master` to `main` ([#143](https://github.com/codeforIATI/iati-datastore/pull/143))
- Compare existing hash of data with new data, in order to massively speed up import ([#157](https://github.com/codeforIATI/iati-datastore/pull/157))
- Switch from TravisCI to Github Actions ([#161](https://github.com/codeforIATI/iati-datastore/pull/161))
- Use CodeforIATI codelists for Reporting Org ([#178](https://github.com/codeforIATI/iati-datastore/pull/178))
- Paginate /api/1/about/deleted/ ([#200](https://github.com/codeforIATI/iati-datastore/pull/200))

### Removed
- Remove the `iati create-database` command (use `iati db upgrade` instead) ([#129](https://github.com/codeforIATI/iati-datastore/pull/129))

### Fixed
- Fix a hierachy bug ([#114](https://github.com/codeforIATI/iati-datastore/pull/114))
- Fix a bug for searching on `recipient-region.text` ([#117](https://github.com/codeforIATI/iati-datastore/pull/117))
- Fix some documentation warnings ([#124](https://github.com/codeforIATI/iati-datastore/pull/124))
- Rationalise `_open_resource` ([#156](https://github.com/codeforIATI/iati-datastore/pull/156))
- Fix integrity constraint error with deleted activities ([#164](https://github.com/codeforIATI/iati-datastore/pull/164))
- Fix `stream=True` for CSV and JSON ([#179](https://github.com/codeforIATI/iati-datastore/pull/179))
- Ensure last_parse_error gets saved ([#201](https://github.com/codeforIATI/iati-datastore/pull/201))
- Fix a bug with /api/1/error/dataset.log, and ensure it uses mimetype `text/plain` ([#210](https://github.com/codeforIATI/iati-datastore/pull/210))
- Fix application context bug ([#217](https://github.com/codeforIATI/iati-datastore/pull/217))

## [1.0.0b1] - 2020-12-10

### Added
- Provide prettier documentation ([#97](https://github.com/codeforIATI/iati-datastore/pull/97); ([#99](https://github.com/codeforIATI/iati-datastore/pull/99)) ([#102](https://github.com/codeforIATI/iati-datastore/pull/102)))
- Add a status API page (/api/1/about/) that provides an accurate indication of status ([#101](https://github.com/codeforIATI/iati-datastore/pull/101))
- Add /meta/filters/ API route ([#105](https://github.com/codeforIATI/iati-datastore/pull/105))
- Add trailing slashes to all routes ([#106](https://github.com/codeforIATI/iati-datastore/pull/106))
- Update README, and add setup instructions ([#107](https://github.com/codeforIATI/iati-datastore/pull/107))

### Changed
- Upgrade to python 3 (and upgrading lots of dependencies) ([#93](https://github.com/codeforIATI/iati-datastore/pull/93))
- Download codelists from https://codelists.codeforiati.org ([#94](https://github.com/codeforIATI/iati-datastore/pull/94))
- Set User-Agent header to "codeforIATI datastore classic" ([#96](https://github.com/codeforIATI/iati-datastore/pull/96))

### Removed
- Remove heroku-specific things ([#103](https://github.com/codeforIATI/iati-datastore/pull/103))

### Fixed
- Updated codelists ([#95](https://github.com/codeforIATI/iati-datastore/pull/95))
- Properly implement CORS ([#100](https://github.com/codeforIATI/iati-datastore/pull/100))


iati_datastore/iatilib/crawler.py
import json
import glob
import datetime
import hashlib
import logging
import traceback
import os

import iatikit
import sqlalchemy as sa
from dateutil.parser import parse as date_parser
from flask import Blueprint
import click

from iatilib import db, parse, rq
from iatilib.model import Dataset, Resource, Activity, Log, DeletedActivity
from iatilib.loghandlers import DatasetMessage as _

log = logging.getLogger("crawler")


manager = Blueprint('crawler', __name__)
manager.cli.short_help = "Crawl IATI registry"


def fetch_dataset_list():
    '''
    Fetches datasets from iatikit and stores them in the DB. Used in update() to update the Flask job queue. Uses CKAN metadata to determine
    if an activity is active or deleted.
    :return:
    '''
    existing_datasets = Dataset.query.all()
    existing_ds_names = set((ds.publisher, ds.name) for ds in existing_datasets)

    package_list = [
        tuple(x[:-5].rsplit('/', 2)[1:])
        for x in glob.glob('__iatikitcache__/registry/metadata/*/*')]
    incoming_ds_names = set(package_list)

    new_datasets = [Dataset(name=n, publisher=p) for p, n
                    in incoming_ds_names - existing_ds_names]
    all_datasets = existing_datasets + new_datasets
    last_seen = iatikit.data().last_updated
    for dataset in all_datasets:
        dataset.last_seen = last_seen

    db.session.add_all(all_datasets)
    db.session.commit()

    deleted_ds_names = existing_ds_names - incoming_ds_names
    if deleted_ds_names:
        delete_datasets([d[1] for d in deleted_ds_names])

    all_datasets = Dataset.query
    return all_datasets


def delete_datasets(datasets):
    deleted_datasets = db.session.query(Dataset).filter(Dataset.name.in_(datasets))

    activities_to_delete = db.session.query(Activity). \
        filter(Activity.resource_url == Resource.url). \
        filter(Resource.dataset_id.in_(datasets))

    now = datetime.datetime.now()
    for a in activities_to_delete:
        db.session.merge(DeletedActivity(
            iati_identifier=a.iati_identifier,
            deletion_date=now
        ))
    db.session.commit()
    deleted = deleted_datasets.delete(synchronize_session='fetch')
    log.info("Deleted {0} datasets".format(deleted))
    return deleted


def fetch_dataset_metadata(dataset):
    fname = '__iatikitcache__/registry/metadata/{0}/{1}.json'.format(
        dataset.publisher, dataset.name)
    with open(fname) as f:
        ds_entity = json.load(f)

    dataset.last_modified = date_parser(
        ds_entity.get(
            'metadata_modified',
            datetime.datetime.now().date().isoformat()))
    new_urls = [resource['url'] for resource
                in ds_entity.get('resources', [])
                if resource['url'] not in dataset.resource_urls]
    dataset.resource_urls.extend(new_urls)

    urls = [resource['url'] for resource
            in ds_entity.get('resources', [])]
    for deleted in set(dataset.resource_urls) - set(urls):
        dataset.resource_urls.remove(deleted)

    dataset.license = ds_entity.get('license_id')
    dataset.is_open = ds_entity.get('isopen', False)
    db.session.add(dataset)
    return dataset


def fetch_resource(dataset):
    '''
    Gets the resource and sets the times of last successful update based on the status code.
    :param resource:
    :return:
    '''
    fname = '__iatikitcache__/registry/data/{0}/{1}.xml'.format(
        dataset.publisher, dataset.name)

    last_updated = iatikit.data().last_updated
    resource = dataset.resources[0]
    resource.last_fetch = last_updated

    if os.path.exists(fname):
        with open(fname, 'rb') as f:
            content = f.read()

        resource.last_status_code = 200
        resource.last_succ = last_updated
        if not resource.document or \
                hash(resource.document) != hash(content):
            resource.document = content
            resource.last_parsed = None
            resource.last_parse_error = None
    else:
        # TODO: this isn't true
        resource.last_status_code = 404

    db.session.add(resource)
    return resource


def check_for_duplicates(activities):
    if activities:
        dup_activity = Activity.query.filter(
                Activity.iati_identifier.in_(
                        a.iati_identifier for a in activities
                )
        )
        with db.session.no_autoflush:
            for db_activity in dup_activity:
                res_activity = next(
                        a for a in activities
                        if a.iati_identifier == db_activity.iati_identifier
                )
                activities.remove(res_activity)
                db.session.expunge(res_activity)
    return activities


def hash(string):
    m = hashlib.md5()
    m.update(string)
    return m.digest()


def parse_activity(new_identifiers, old_xml, resource):
    for activity in parse.document_from_bytes(resource.document, resource):
        activity.resource = resource

        if activity.iati_identifier not in new_identifiers:
            new_identifiers.add(activity.iati_identifier)
            try:
                if hash(activity.raw_xml.encode('utf-8')) == old_xml[activity.iati_identifier][1]:
                    activity.last_change_datetime = old_xml[activity.iati_identifier][0]
                else:
                    activity.last_change_datetime = datetime.datetime.now()
            except KeyError:
                activity.last_change_datetime = datetime.datetime.now()
            db.session.add(activity)
            check_for_duplicates([activity])
        else:
            parse.log.warn(
                    _("Duplicate identifier {0} in same resource document".format(
                            activity.iati_identifier),
                      logger='activity_importer', dataset=resource.dataset_id, resource=resource.url),
                    exc_info=''
            )

        db.session.flush()
    db.session.commit()


def parse_resource(resource):
    db.session.add(resource)
    current = Activity.query.filter_by(resource_url=resource.url)
    current_identifiers = set([i.iati_identifier for i in current.all()])

    # obtains the iati-identifier, last-updated datetime, and a hash of the existing xml associated with
    # every activity associated with the current url.
    old_xml = dict([(i[0], (i[1], hash(i[2].encode('utf-8')))) for i in db.session.query(
            Activity.iati_identifier, Activity.last_change_datetime,
            Activity.raw_xml).filter_by(resource_url=resource.url)])

    db.session.query(Activity).filter_by(resource_url=resource.url).delete()
    new_identifiers = set()
    parse_activity(new_identifiers, old_xml, resource)

    resource.version = parse.document_metadata(resource.document)

    # add any identifiers that are no longer present to deleted_activity table
    diff = current_identifiers - new_identifiers
    now = datetime.datetime.utcnow()
    deleted = [
        DeletedActivity(iati_identifier=deleted_activity, deletion_date=now)
        for deleted_activity in diff]
    if deleted:
        db.session.add_all(deleted)

    # remove any new identifiers from the deleted_activity table
    if new_identifiers:
        db.session.query(DeletedActivity) \
            .filter(DeletedActivity.iati_identifier.in_(new_identifiers)) \
            .delete(synchronize_session="fetch")

    log.info(
            "Parsed %d activities from %s",
            resource.activities.count(),
            resource.url)
    resource.last_parsed = now
    return resource  # , new_identifiers


def update_activities(dataset_name):
    '''
    Parses and stores the raw XML associated with a resource [see parse_resource()], or logs the invalid resource
    :param resource_url:
    :return:
    '''
    # clear up previous job queue log errors
    db.session.query(Log).filter(sa.and_(
            Log.logger == 'job iatilib.crawler.update_activities',
            Log.resource == dataset_name,
    )).delete(synchronize_session=False)
    db.session.commit()

    dataset = Dataset.query.get(dataset_name)
    resource = dataset.resources[0]
    try:
        db.session.query(Log).filter(sa.and_(
                Log.logger.in_(
                        ['activity_importer', 'failed_activity', 'xml_parser']),
                Log.resource == dataset_name,
        )).delete(synchronize_session=False)
        parse_resource(resource)
        db.session.commit()
    except parse.ParserError as exc:
        db.session.rollback()
        resource.last_parse_error = str(exc)
        db.session.add(resource)
        db.session.add(Log(
                dataset=resource.dataset_id,
                resource=resource.url,
                logger="xml_parser",
                msg="Failed to parse XML file {0} error was".format(dataset_name, exc),
                level="error",
                trace=traceback.format_exc(),
                created_at=datetime.datetime.now()
        ))
        db.session.commit()


def update_dataset(dataset_name):
    '''
    Takes the dataset name and determines whether or not an update is needed based on whether or not the last
    successful update detail exits, and whether or not it last updated since the contained data was updated.
    :param dataset_name:
    :return:
    '''
    # clear up previous job queue log errors
    db.session.query(Log).filter(sa.and_(
            Log.logger == 'job iatilib.crawler.update_dataset',
            Log.dataset == dataset_name,
    )).delete(synchronize_session=False)
    db.session.commit()

    queue = rq.get_queue()
    dataset = Dataset.query.get(dataset_name)

    fetch_dataset_metadata(dataset)
    try:
        db.session.commit()
    except sa.exc.IntegrityError as exc:
        db.session.rollback()
        # the resource can't be added, so we should
        # give up.
        db.session.add(Log(
            dataset=dataset_name,
            resource=None,
            logger="update_dataset",
            msg="Failed to update dataset {0}, error was".format(dataset_name, exc),
            level="error",
            trace=traceback.format_exc(),
            created_at=datetime.datetime.now()
        ))
        db.session.commit()
        return

    resource = fetch_resource(dataset)
    db.session.commit()

    if resource.last_status_code == 200 and not resource.last_parsed:
        queue.enqueue(
            update_activities, args=(dataset_name,),
            result_ttl=0, job_timeout=100000)


def status_line(msg, filt, tot):
    total_count = tot.count()
    filtered_count = filt.count()
    try:
        ratio = 1.0 * filtered_count / total_count
    except ZeroDivisionError:
        ratio = 0.0
    return "{filt_c:4d}/{tot_c:4d} ({pct:6.2%}) {msg}".format(
            filt_c=filtered_count,
            tot_c=total_count,
            pct=ratio,
            msg=msg
    )


@manager.cli.command('status')
def status_cmd():
    """Show status of current jobs"""
    print("%d jobs on queue" % rq.get_queue().count)

    print(status_line(
            "datasets have no metadata",
            Dataset.query.filter_by(last_modified=None),
            Dataset.query,
    ))

    print(status_line(
            "datasets not seen in the last day",
            Dataset.query.filter(Dataset.last_seen <
                                 (datetime.datetime.utcnow() - datetime.timedelta(days=1))),
            Dataset.query,
    ))

    print(status_line(
            "resources have had no attempt to fetch",
            Resource.query.outerjoin(Dataset).filter(
                    Resource.last_fetch == None),
            Resource.query,
    ))

    print(status_line(
            "resources not successfully fetched",
            Resource.query.outerjoin(Dataset).filter(
                    Resource.last_succ == None),
            Resource.query,
    ))

    print(status_line(
            "resources not fetched since modification",
            Resource.query.outerjoin(Dataset).filter(
                    sa.or_(
                            Resource.last_succ == None,
                            Resource.last_succ < Dataset.last_modified)),
            Resource.query,
    ))

    print(status_line(
            "resources not parsed since mod",
            Resource.query.outerjoin(Dataset).filter(
                    sa.or_(
                            Resource.last_succ == None,
                            Resource.last_parsed < Dataset.last_modified)),
            Resource.query,
    ))

    print(status_line(
          "resources have no activites",
          db.session.query(Resource.url).outerjoin(Activity)
          .group_by(Resource.url)
          .having(sa.func.count(Activity.iati_identifier) == 0),
          Resource.query))

    print("")

    total_activities = Activity.query.count()
    # out of date activitiy was created < resource last_parsed
    total_activities_fetched = Activity.query.join(Resource).filter(
        Activity.created < Resource.last_parsed).count()
    try:
        ratio = 1.0 * total_activities_fetched / total_activities
    except ZeroDivisionError:
        ratio = 0.0
    print("{nofetched_c}/{res_c} ({pct:6.2%}) activities out of date".format(
            nofetched_c=total_activities_fetched,
            res_c=total_activities,
            pct=ratio
    ))


@manager.cli.command('download')
def download_cmd():
    """
    Download all IATI data from IATI Data Dump.
    """
    iatikit.download.data()


def download_and_update():
    iatikit.download.data()
    update_registry()


@manager.cli.command('fetch-dataset-list')
def fetch_dataset_list_cmd():
    """
    Fetches dataset metadata from existing IATI Data Dump cache.
    """
    fetch_dataset_list()


@manager.cli.command('download-and-update')
def download_and_update_cmd():
    """
    Enqueue a download of all IATI data from
    IATI Data Dump, and then start an update.
    """
    queue = rq.get_queue()
    print("Enqueuing a download from IATI Data Dump")
    queue.enqueue(
        download_and_update,
        result_ttl=0, job_timeout=100000)


def update_registry():
    queue = rq.get_queue()
    datasets = fetch_dataset_list()
    print("Enqueuing %d datasets for update" % datasets.count())
    for dataset in datasets:
        queue.enqueue(update_dataset, args=(dataset.name,), result_ttl=0)


@click.option('--dataset', 'dataset', type=str,
              help="update a single dataset")
@manager.cli.command('update')
def update_cmd(dataset=None):
    """
    Step through downloaded datasets, adding them to the dataset table, and then adding an update command to
    the Flask job queue. See update_registry, then update_dataset for next actions.
    """
    queue = rq.get_queue()

    if dataset is not None:
        print("Enqueuing {0} for update".format(dataset))
        queue.enqueue(update_dataset, args=(dataset,), result_ttl=0)
    else:
        print("Enqueuing a full registry update")
        queue.enqueue(update_registry, result_ttl=0)
